README.md
# misc-scripts
These are random scripts I wrote while at Korelogic. Names with an
asterisk (*) are in an unknown state.

## Scripts
- bashpack.pl - packs sourced scripts into one file
- host-vars.example - example file to go along with mon-hosts
- ipt-count-trace* - run a command twice and see the iptables counter
difference between the runs
- mon-hosts-packed - the packed up version of mon-hosts
- mon-hosts.sh - a script to monitor/run a command on list(s) of hosts
- sshs.sh* - simple script to ssh to a server and determine an appropriate
terminal and attach to the default screen session
- tmux-start - run a command with different parameters in separate tmux
windows (like starting ssh to a bunch of hosts that then uses sshs to open
your screen session there?)

## Copyright and Thanks

Author: Shawn Wilson <ag4ve.us@gmail.com>

GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007 http://www.gnu.org/licenses/gpl-3.0.html

I thank Korelogic for their support on this project.



ChangeLog
2015-12-12 11:26  swilson

	* mon-hosts.sign (1.9): Update signature.

2015-12-12 11:26  swilson

	* mon-hosts (1.7): Remove path.

2015-12-12 11:23  swilson

	* mon-hosts.sign (1.8): Update signature.

2015-12-12 11:23  swilson

	* mon-hosts (1.6): Upstream changes.

2015-12-12 11:22  swilson

	* mon-hosts.sh.sign (1.7): Update signature.

2015-12-12 11:22  swilson

	* mon-hosts.sh (1.46): Remove double space from help text.

2015-12-12 08:17  swilson

	* bashpack.tgz.sign (1.1): Update signature.

2015-12-12 08:17  swilson

	* bashpack.tgz (1.1): Initial bashpack archive.

2015-12-12 04:45  swilson

	* mon-hosts.sign (1.7): Update signature.

2015-12-12 04:45  swilson

	* mon-hosts (1.5): Actually pack the script.

2015-12-12 04:04  swilson

	* mon-hosts.sign (1.6): Update signature.

2015-12-12 04:04  swilson

	* mon-hosts (1.4): Upstream changes.

2015-12-12 04:03  swilson

	* mon-hosts.sh.sign (1.6): Update signature.

2015-12-12 04:03  swilson

	* mon-hosts.sh (1.45): Allow hosts as IPs, Make an exported option
	  to take everything from a command line, Add examples.

2015-12-11 23:43  swilson

	* mon-hosts.sh.sign (1.5): Update signature.

2015-12-11 23:43  swilson

	* mon-hosts.sh (1.44): Add a date format option.

2015-12-11 23:31  swilson

	* mon-hosts.sh (1.43): Make the change file work and display less
	  wordy output. Add comments and debugging.

2015-12-11 23:44  swilson

	* mon-hosts (1.3): Upstream changes.

2015-12-11 23:44  swilson

	* mon-hosts.sign (1.4), mon-hosts.sign (1.5): Update signature.

2015-12-11 09:18  swilson

	* mon-hosts.sh (1.42): Specify null config file so that -l foo does
	  not confuse the test.

2015-12-11 09:16  swilson

	* mon-hosts.sh (1.41): Include text and example file from the web
	  site so that use is still obvious without site text.

2015-12-11 09:07  swilson

	* mon-hosts.sh (1.40): Allow a command with options - only test the
	  first word is a command. Add a comment for an easy way to test.

2015-11-20 15:23  swilson

	* mon-hosts.sign (1.3): Update signature.

2015-11-20 15:23  swilson

	* mon-hosts (1.2): Update with upstream changes.

2015-11-20 15:15  swilson

	* mon-hosts.sh.sign (1.4): Update signature.

2015-11-20 15:15  swilson

	* mon-hosts.sh (1.39): Add host list format and defaults to help.

2015-11-20 15:03  swilson

	* mon-hosts.sh (1.38): Add comments. make eval_host() so that
	  variables can be scoped and move code out of the main for loop
	  into that. Allow for <host>=<command> for a specific command to
	  be run on that host/group in addition to having a global -a
	  option for those commands. Add debug for bell.

2015-11-20 11:57  swilson

	* mon-hosts.sh (1.37): Add bell option.

2015-11-08 00:31  swilson

	* ipt-count-trace.sign (1.2), mon-hosts.sign (1.2): Update
	  signature.

2015-11-06 17:14  swilson

	* mon-hosts.sh (1.36): Change env so that it has more chance of
	  running on OSX - probably still would not because of system bash
	  3.

2015-11-06 17:12  swilson

	* bashpack.pl (1.8): Allow for spaces as an interpreter can be a
	  command such as 'env bash'.

2015-11-06 17:10  swilson

	* bashpack.pl (1.7): Put a space between shabang and interpreter.

2015-11-02 04:47  swilson

	* tmux-start.sh.sign (1.3), mon-hosts.sh.sign (1.3): Update
	  signature.

2015-11-01 15:16  swilson

	* tmux-start.sh (1.16): Do not display the tmux check.

2015-10-31 06:47  swilson

	* mon-hosts.sh (1.35): Add date to state change line.

2015-10-31 06:37  swilson

	* mon-hosts.sh (1.34): Add changefile code.

2015-10-30 08:24  swilson

	* mon-hosts.sh (1.33): Add/correct documentation, do not
	  increase/decrease min/max if they are already above the
	  threshold.

2015-10-29 22:19  swilson

	* mon-hosts.sh (1.32): Capitalize first letter in help message. Add
	  a debug file option. Add an option for a future feature.

2015-10-29 21:16  swilson

	* mon-hosts.sh (1.31): Fix documentation and retries - should only
	  retry once by default.

2015-10-29 21:02  swilson

	* mon-hosts.sh (1.30): Default minwidth now that control sequences
	  are not expected.

2015-10-29 20:54  swilson

	* mon-hosts.sh (1.29): Remove control characters.

2015-10-29 20:51  swilson

	* mon-hosts.sh (1.28): Account for surrounding brackets.

2015-10-29 20:37  swilson

	* mon-hosts.sh (1.27): For some reason eval does not seem to return
	  the exit code to while to end - use for counter and if instead -
	  cleaner code anyway. Check runs before evaluating it.

2015-10-29 08:49  swilson

	* mon-hosts.sh (1.26): Add an option to define the number of runs.

2015-10-29 06:45  swilson

	* mon-hosts.sh (1.25): Add comments and better formatting and
	  change variable names to lower case (since that is modern bash
	  convention) and try to make function variables local. Move
	  rerun_cmd() parameter declaration to one place. Add debugging for
	  new command line options. Check for the command variable since it
	  might not be ssh.

2015-10-29 05:05  swilson

	* mon-hosts.sh (1.24): Add lots of math to determine padding to
	  make sure that everything lines up when printed.

2015-10-29 03:42  swilson

	* mon-hosts.sh (1.23): Should be adding a space - not removing one.
	  Fields align now. Add -X option to set a minimum status width.
	  Make sure options that should be numbers are numbers.

2015-10-28 09:58  swilson

	* mon-hosts.sh.sign (1.2): Update signature.

2015-10-28 06:03  swilson

	* mon-hosts.sh (1.22): Try to prevent erroring commands from
	  showing up (like if the server can not be reached).

2015-10-27 22:51  swilson

	* mon-hosts.sh (1.21): Easier to read.

2015-10-27 22:48  swilson

	* mon-hosts.sh (1.20): Syntax err.

2015-10-27 22:42  swilson

	* mon-hosts.sh (1.19): Account for spaces when printing.

2015-10-25 20:05  swilson

	* mon-hosts.sh.sign (1.1): Add mon-hosts signature.

2015-10-25 20:04  swilson

	* mon-hosts.sh (1.18): Modify help message to warn about the
	  implications of the sourced config file.

2015-10-25 19:59  swilson


bashpack.pl
#!/usr/bin/env perl

use strict;
use warnings;

my @paths = split(':', $ENV{PATH});
push(@paths, ".");

my @files;
push @files, @ARGV;

my $data = {};
my $count = 0;
my $interp;

while (scalar(@files))
{
  my $file = shift(@files);
  # File has been seen
  if (scalar($data) eq 'HASH' and exists($data->{$file}))
  {
    $data->{$file}{count} = $count++;
    next;
  }
  # No file to see
  elsif (! -e $file)
  {
    next;
  }

  open (my $fh, "<", "$file");
  my $buf;
  while (my $line = <$fh>)
  {
    chomp $line;
    # Replace file strings (not variables - which can be dynamic) that are 
    # sourced with the file string
    # TODO spaces are allowed in a file path - are not accepted here
    if ($line =~ /^ *source *([^ \$]+) *$/)
    {
      my $srcstring = $1;
      # Make sure quotes match and strip them
      $srcstring =~ s/^("[^"]+"|'[^']+'|[^'"]+)$/$1/;
      $srcstring =~ s/"'//g;
      if (not length($srcstring))
      {
        warn "Bad source line [$line]\n";
        $buf .= "$line\n";
      }
      else
      {
        my $fullpath;
        # confirm absolute path
        if ($srcstring =~ /\//)
        {
          $fullpath = $srcstring if (-e "$srcstring");
        }
        # search through paths and find the file if it's there
        else
        {
          foreach my $path (@paths)
          {
            if (-e "$path/$srcstring")
            {
              $fullpath = "$path/$srcstring";
              last;
            }
          }
        }
        if (not length($fullpath))
        {
          warn "Source file not found [$line]\n";
          $buf .= "$line\n";
        }
        else
        {
          push @files, $fullpath;
          $buf .= "# Sourced $srcstring\n";
        }
      }
    }
    # Handle shabang
    elsif ($. eq 1 and $line =~ /^#! ?([a-zA-Z0-9\/.,+_ -]+)/)
    {
      if (not defined($interp))
      {
        $interp = $1;
      }
    }
    else
    {
      $buf .= "$line\n";
    }
  }
  $data->{$file}{buf} = $buf;

  $data->{$file}{count} = $count++;
}

print "#! " . $interp . "\n" if (defined($interp));

# Reverse sort file buffers in our hash and print them to stdout
for my $file 
  (
    sort {$data->{$b}{count} <=> $data->{$a}{count}}
    keys %$data
  )
{
  print "# Included $file\n";
  print $data->{$file}{buf};
  print "\n";
}


tmux-start.sh
#!/usr/bin/env bash

# $Id: tmux-start.sh,v 1.16 2015/11/01 22:16:34 swilson Exp $

# Author: Shawn Wilson <swilson@korelogic.com>
# Copyright 2015 Korelogic Inc. https://www.korelogic.com
# GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007
# http://www.gnu.org/licenses/gpl-3.0.html

# basename
script="${0##*/}"

IFS="." read -a bash_version <<<"${BASH_VERSION}"

die()
{
  echo -ne "$@" >&2
  exit 1
}

warn()
{
  echo "WARN: $@\n" >&2
}

debug()
{
  if [ -n "$DEBUG" ] ; then
    warn "$@"
  fi
}

opthelp()
{
  die "$@\n" \
    "$script [OPTIONS]\n" \
    "   -a    string to append to the command for each variable\n" \
    "   -c    command to run\n" \
    "   -f    force (restarts session with the same name)\n" \
    "   -n    name of the array to use\n" \
    "   -p    prefix to insert into the tmux session name\n" \
    "   -P    Pause between commands\n" \
    "   -r    Max retries per command\n" \
    "   -s    sourced config file with array of variables (warning: " \
      "contains code)\n" \
    "   -S    interactive shell (eg: bash --norc --noprofile)\n" \
    "   -x    run each element as a full command\n" \
    "   -d    debuging\n" \
    "   -h    this message\n"
}

session_test ()
{
  local session="$@"

  tmux list-window -t "$session" 2>/dev/null | wc -l
}

# Will be exported to every session that is started
_rerun_cmd ()
{
  local cmd="$1"
  if [[ -z "$cmd" ]] ; then
    return
  fi

  local max="$2"
  # Max needs to be defined just in case the command will never succeed
  if [[ -z "$max" ]] ; then
    max=10
  fi

  local sleep="$3"
  if [[ -z "$sleep" ]] ; then
    sleep=1
  fi

  local count=0
  while eval $cmd ; do
    if ((count >= max)) ; then
      break
    fi
    sleep "$sleep";
  done
}

config=~/.config/host-vars.sh

# Catch no options
if [[ "$#" -eq 0 ]] ; then
  opthelp
fi

# Default values
pause=0

while getopts ":a:n:c:fP:p:r:S:s:xdh" opt; do
  case "$opt" in
    # Append string
    a)
      append="${OPTARG}"
      ;;
    # Command
    c)
      cmd="${OPTARG}"
      ;;
    # Force a kill and respawn of a session with the same name
    f)
      force=1
      ;;
    # Variable name
    n)
      name="${OPTARG}"
      ;;
    # Pause value
    P)
      pause="${OPTARG}"
      ;;
    # Tmux session name prefix
    p)
      prefix="${OPTARG}"
      ;;
    # Max retries
    r)
      retries="${OPTARG}"
      ;;
    # Shell
    S)
      shell="${OPTARG}"
      ;;
    # Config file
    s)
      [[ -e "${OPTARG}" ]] || die "Config file [${OPTARG}] not found."
      config="${OPTARG}"
      ;;
    # No command eval
    x)
      nocmd=1
      ;;
    # Debug
    d)
      debug=1
      ;;
    # Help
    h)
      opthelp
      ;;
    *)
      opthelp
      ;;
  esac
done
shift $((OPTIND-1))

# Make sure we know if nothing is specified to be run
if [[ -z "$cmd" && -z "$nocmd" ]] ; then
  die "The -x option must be specified if not giving a command.\n"
fi

# Check proper command
if ! ( [[ -x "${cmd%% *}" ]] || type "${cmd%% *}" >/dev/null 2>&1 ) ; then
  die "Not a command [${cmd%% *}]\n"
else
  read -r acmd aparams <<<"${cmd}"
  fullcmd="$(type -p $acmd)"
  if [[ -n "$fullcmd" ]] ; then
    cmd="$fullcmd $aparams"
  else
    cmd="$acmd $aparams"
  fi
fi

# Sanity checks
if [ "${bash_version[0]}" -lt 4 ] ; then
  die "Bash version less than 4 is not supported\n"
fi
if ! type tmux >/dev/null 2>&1 ; then
  die "tmux does not exist\n"
fi

source "$config" 2>/dev/null

unset var
declare -n var="$name"

if [[ -z "$var" ]] ; then
  die "$name not defined in config [$config]\n"
else
  debug "Elements: ${var[@]}"
fi

if [[ -z "$shell" ]] ; then
  case "${SHELL##*/}" in
    bash)
      shell="bash --norc --noprofile"
      ;;
    zsh)
      shell="zsh -f -i"
      ;;
  esac
fi

offset=0
for ((count = 0 ; count < 10 ; count++)) ; do
  session="${prefix}${name}${count}"

  # Increment if the session name is already in use
  if [[ $(session_test "$session") -gt 0 ]] ; then
    if [[ -n "$force" ]] ; then
      tmux kill-session -t "$session"
    else
      continue
    fi
  fi

  # Create a new session - temporarily unset TMUX so it doesn't complain
  # if already inside tmux - -d makes this DWIM
  TMUX= tmux -2 new-session -d -s "$session"

  while [ $(session_test "$session") -eq 0 ] ; do
    sleep 1
  done

  # Make sure we only start 10 windows per session so meta-[0-9] work for
  # all windows
  if ((offset + 10 > ${#var[@]})) ; then
    max=$((${#var[@]} - $offset))
    end=1
  else
    max=10
  fi

  for ((num = 0 ; num < "$max" ; num++)) ; do
    # Window name
    unit="${var[${offset} + ${num}]}${append}"

    debug "Window name [$unit] Session [${session}:${num}]"

    # Start off with a new window
    if [[ "$num" -gt 0 ]] ; then
      tmux new-window -t "${session}:${num}" -n "$unit" -- "$shell"
    fi

    # Declare a private function to give commands (like ssh that might get
    # bogged down) a chance to start
    tmux send-keys -t "${session}:${num}" " $(declare -f _rerun_cmd)" C-m
    tmux send-keys -t "${session}:${num}" " clear" C-m
    tmux send-keys -t "${session}:${num}" "_rerun_cmd \"${cmd} ${unit}\" \"$retries\" \"$pause\"" C-m
  done

  # end when the array "var" is done
  if [[ -n "$end" ]] ; then
    break
  fi

  ((offset+=10))
done



mon-hosts.sh
#!/usr/bin/env bash

# $Id: mon-hosts.sh,v 1.46 2015/12/12 18:22:24 swilson Exp $

# Author: Shawn Wilson <swilson@korelogic.com>
# Copyright 2015 Korelogic Inc. https://www.korelogic.com
# GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007
# http://www.gnu.org/licenses/gpl-3.0.html

# A bash script to show the status of a group or groups of hosts based on a
# config file and command-line parameters. An command (based on the example
# config) might look like: 
# mon-hosts -l foo,bar", or "mon-hosts -l everything -a 'uname -r' -R".

# Example file:
# #!/bin/bash -
# 
# # create a list named "foo"
# foo=()
# foo+=(perlish)
# foo+=(tester)
# 
# # create another list named "bar"
# bar=()
# bar+=(nibble)
# bar+=(osx)
# 
# # a list containing all elements from foo and bar
# everything=()
# everything+=(${foo[*]})
# everything+=(${bar[*]})

# Test by creating/removing a local file (foo) to show foo as up/down:
# ./mon-hosts -c 'test -e' -a '' -s /dev/null -l foo
# Or to check if a host is up by pinging:
# ./mon-hosts.sh -c 'ping -c 1' -a '>/dev/null 2>&1' -s /dev/null -l 10.0.0.1
# Command line IP ping scan (takes a while - don't do this - the use of ping 
# is for interoperability with all Unices):
# ips=(10.0.0.{1..20}) && export exported=$(declare -p ips) && ./mon-hosts.sh -c 'ping -c 1 -w 1 -W 1' -a '>/dev/null 2>&1' -s /dev/null -l ips -e

# Default values
append="echo -n"
cmd="ssh"
up="*"
down="X"
pause=5
sleep=1
retries=1
usereturn=0
config=~/.config/host-vars.sh
cols=1
minwidth=15
datefmt="%c"

set -o nounset
# set -eix

# Make sure we're using a sane locale
export LANG=C

# basename
script="${0##*/}"

# bash-libs needs to be in your path - also see bashpack
source msg-handling.sh

# Try to rerun a command until it succeeds or a limit it hit
rerun_cmd ()
{
  debug 6 "rerun_cmd options [$@]"

  unset rerun_cmd_ret
  declare -ga rerun_cmd_ret

  declare cmd="${1:-}"
  declare host="${2:-}"
  declare append="${3:-}"
  declare max="${4:-1}"
  declare sleep="${5:-0}"

  if [[ -z "${cmd:-}" || -z "${host:-}" ]] ; then
    return
  fi

  declare fullcmd="$cmd $host $append 2>/dev/null"
  declare origup="$up"

  # If we're returning command data from the host
  # $up is the return data from the command in this case
  if (( usereturn != 0 )) ; then
    declare up=""
    fullcmd="up=\"\$($fullcmd)\""
  fi
  debug 4 "rerun_cmd fullcmd [$fullcmd]"

  for (( count = 0 ; count < max ; count++ )) ; do
    if eval $fullcmd ; then
      break
    fi
    sleep "${sleep:-0}"
  done
  debug 5 "rerun_cmd count [$count] max [$max] up [$up] down [$down]"

  # Determine status
  if (( count < max )) ; then
    # Make sure there's something to return
    if [[ -z "${up:-}" ]] ; then
      up="$origup"
    fi

    rerun_cmd_ret=("$host" "[$up]")
  else
    rerun_cmd_ret=("$host" "[$down]")
    if [[ -n "${bell:-}" ]] ; then
      echo -ne "\a"
    fi
  fi

  debug 4 "rerun_cmd_ret [${rerun_cmd_ret[@]:-}]"
}

# Format host data
host_print()
{
  debug 6 "host_print options [$@]"

  if [[ -z "${1:-}" ]] ; then
    return
  fi

  declare -n status="${1:-}"
  declare group="${2:-}"
  declare cols="${3:-1}"

  # Calculate widths 

  declare maxhost=0
  declare maxupdown=0

  # Determine the maximum width of hosts and status value
  for ((count = 0 ; count < "${#status[@]:-}" ; count+=2)) ; do
    hostlength="${#status[$count]}"
    debug 7 "host [${status[$count]}] length [$hostlength]"
    if (( hostlength > maxhost )) ; then
      maxhost="$hostlength"
    fi
    if [[ -n "${usereturn:-}" ]] ; then
      updownlength="${#status[$count +1]}"
      debug 7 "updown [${status[$count +1]}] length [$updownlength]"
      if (( updownlength > maxupdown )) ; then
        maxupdown="$updownlength"
      fi
    fi
  done
  debug 5 "host_print maxhost length [$maxhost]"

  for i in "${#up}" "${#down}" ; do
    if (( i > maxupdown )) ; then
      maxupdown="$i"
    fi
  done
  debug 5 "host_print maxupdown length [$maxupdown]"

  # Make sure lengths aren't greater than the maxes or mins and that the
  # max/min value wouldn't be greater or less than their current value
  # if they are host  gets 75% of the width and status gets 25%
  if [[ -n "${maxwidth:-}" ]] && (( (maxhost + maxupdown) > maxwidth )) ; then
    declare tmaxhost=$(( (maxwidth / 4) * 3))
    if (( tmaxhost < maxhost )) ; then
      maxhost="${tmaxhost:-}"
    fi
    declare tmaxupdown=$(( (maxwidth / 4) * 1))
    if (( tmaxupdown < maxupdown )) ; then
      maxupdown="${tmaxupdown:-}"
    fi
    debug 5 "host_print length max reset: maxhost [$maxhost] maxupdown [$maxupdown]"
  fi
  if [[ -n "${minwidth:-}" ]] && (( (maxhost + maxupdown) < minwidth )) ; then
    declare tmaxhost=$(( (minwidth / 4) * 3))
    if (( tmaxhost > maxhost )) ; then
      maxhost="${tmaxhost:-}"
    fi
    declare tmaxupdown=$(( (minwidth / 4) * 1))
    if (( tmaxupdown > maxupdown )) ; then
      maxupdown="${tmaxupdown:-}"
    fi
    debug 5 "host_print length min reset: maxhost [$maxhost] maxupdown [$maxupdown]"
  fi
  if [[ -n "${minstatus:-}" ]] && (( maxupdown < minstatus )) ; then
    maxupdown="$minstatus"
    debug 5 "host_print length minstatus reset: maxupdown [$maxupdown]"
  fi

  # Calculate the number of columns that fit on a line
  # <allowed colunms> = 
  # <terminal space> / <hostname space> + <status space> + <added whitespace>
  declare colnum=$(( $cols / ($maxhost + $maxupdown +3) ))
  if (( colnum < 1 )) ; then
    colnum=1
  fi
  debug 4 "host_print colnum [$colnum] cols [$cols]"

  # Print a group if it's defined
  if [[ -n "${group:-}" ]] ; then
    host_print_ret="${host_print_ret}${group}\n "
  fi

  # Loop through data and properly format it
  declare newline=1
  for ((count = 0 ; count < "${#status[@]}" ; count+=2)) ; do
    # Determine if we need a line break
    if (( newline == 0 && ((count / 2) % colnum) == 0 )) ; then
      host_print_ret="$host_print_ret\n "
      newline=1
    fi

    # Set pad if required (one space is applied by default): 
    # <max field number> - <length of variable> + <space>
    declare hostpad=$(( maxhost - ${#status[$count]} +1 ))
    if (( hostpad < 0 )) ; then
      hostpad=1
    fi
    declare updownpad=$(( maxupdown - ${#status[$count +1]} +1 ))
    if (( updownpad < 0 )) ; then
      updownpad=1
    fi

    # Append formatted line part to return variable
    printf -v host_print_ret \
      "%s%.${maxhost}s%-${hostpad}s%.${maxupdown}b%-${updownpad}s" \
      "$host_print_ret" \
      "${status[$count]}" \
      " " \
      "${status[$count +1]}" \
      " "

    newline=0
  done

  # A newline won't have been appended
  host_print_ret="$host_print_ret\n"

  # Only add a separating line if hosts were printed
  if [[ "$group\n" != "$host_print_ret" ]] ; then
    host_print_ret="${host_print_ret}${line}\n"
  fi
}

# Do work for each host or group of hosts
eval_host()
{
  declare hoststr="${1:-}"
  declare hostvar
  declare hostcmd
  IFS="=" read hostvar hostappend <<<"${hoststr:-}"
  # Make sure hostvar is defined and a valid variable name
  if [[ -n "${hostvar:-}" && $hostvar =~ ^[a-zA-Z_][a-zA-Z_0-9]*$ ]] ; then
    declare -n var="$hostvar"
  fi
  # If there's no individual <hosts>=<cmd> use the default append command
  if [[ -z "${hostappend:-}" ]] ; then
    hostappend="${append:-}"
  fi
  
  debug 5 "hostvar [$hostvar] var [" $(declare -p var 2>/dev/null) "]"
  # The hostvar is not in the config so hostvar should be a hostname
  if [[ -z "${var:-}" ]] ; then
    rerun_cmd "$cmd" "$hostvar" "$hostappend" "${retries:-}" "${sleep:-}"
    host_print "rerun_cmd_ret" "$cols"
    if [[ -n "${changefile:-}" ]] ; then
        data+=("${rerun_cmd_ret[@]:-}")
    fi
    unset rerun_cmd_ret
  # This should be the default - hostvar is defined
  else
    declare -a statushosts
    for host in "${var[@]}" ; do
      rerun_cmd "$cmd" "$host" "$hostappend" "${retries:-}" "${sleep:-}"
      statushosts+=("${rerun_cmd_ret[0]:-}")
      statushosts+=("${rerun_cmd_ret[1]:-}")
    done
    host_print "statushosts" "$hostvar" "$cols"
    if [[ -n "${changefile:-}" ]] ; then
        data+=("${statushosts[@]:-}")
    fi
    unset statushosts
  fi
}

opthelp()
{
  die "$@\n" \
    "$script [OPTIONS]\n" \
    "   -a    String to append to the command (default: echo -n)\n" \
    "   -b    Audible bell when hosts are down (default: no)\n" \
    "   -c    Command to run (default: ssh)\n" \
    "   -C    Do not clear between runs (default: no)\n" \
    "   -e    Eval an \$exported variable (default: no)\n" \
    "   -F    Date format (default: %c)\n" \
    "   -l    Comma separated list of hosts or arrays of hosts and optional" \
      "individual commands (required:" \
      "host0[=<cmd>],host1[=<cmd>],...hostn[=<cmd>])\n" \
    "   -n    Number of runs (default: infinite)\n" \
    "   -m    Down mark (default: [X])\n" \
    "   -M    Up mark (default: [*])\n" \
    "   -p    Pause between runs (default: 5)\n" \
    "   -P    Pause between retries (default: 1)\n" \
    "   -r    Max retries per command (default: 1)\n" \
    "   -R    Use the return value instead of the up value (default: no)\n" \
    "   -s    Sourced config file with array of variables (warning:" \
      "contains code, default: ~/.config/host-vars.sh)\n" \
    "   -S    Status change file (default: none)\n" \
    "   -w    Max total column (name and status) width (default: none)\n" \
    "   -W    Min total column (name and status) width (default: 15)\n" \
    "   -X    Min status width (default: none)\n" \
    "   -d    Debug level (default: 0)\n" \
    "   -D    Debug file (default: none)\n" \
    "   -h    This message\n"
}

while getopts ":a:bc:CeF:l:m:M:n:p:P:r:Rs:S:w:W:X:d:D:h" opt; do
  case "$opt" in
    # Append string
    a)
      append="${OPTARG}"
      ;;
    # Bell
    b)
      bell=1
      ;;
    # Command
    c)
      cmd="${OPTARG}"
      ;;
    # Do not clear
    C)
      noclear=1
      ;;
    # Exported variable name
    e)
      exports=1
      ;;
    # Date format
    F)
      datefmt="${OPTARG}"
      ;;
    # Host list
    l)
      IFS="," read -a hosts <<<"${OPTARG}"
      ;;
    # Down mark
    m)
      down="${OPTARG}"
      ;;
    # Up mark
    M)
      up="${OPTARG}"
      ;;
    # Number of runs
    n)
      [[ "${OPTARG}" == [0-9]* ]] || die "-n should be an integer\n"
      runs="${OPTARG}"
      ;;
    # Pause time
    p)
      [[ "${OPTARG}" == [0-9]* ]] || die "-p should be an integer\n"
      pause="${OPTARG}"
      ;;
    # Max retries
    r)
      [[ "${OPTARG}" == [0-9]* ]] || die "-r should be an integer\n"
      retries="${OPTARG}"
      ;;
    # Use return value
    R)
      usereturn=1
      ;;
    # Config file
    s)
      [[ -e "${OPTARG}" ]] || die "Config file [${OPTARG}] not found\n"
      config="${OPTARG}"
      ;;
    # Status change file
    S)
      [[ ( -n "${OPTARG%/*}" && -w "${OPTARG%/*}/." ) || -w "./." ]] || die \
        "Can not write change file [${OPTARG}]\n"
      changefile="${OPTARG}"
      ;;
    # Max cel width
    w)
      [[ "${OPTARG}" == [0-9]* ]] || die "-w should be an integer\n"
      maxwidth="${OPTARG}"
      ;;
    # Min col width
    W)
      [[ "${OPTARG}" == [0-9]* ]] || die "-W should be an integer\n"
      minwidth="${OPTARG}"
      ;;
    # Min status width
    X)
      [[ "${OPTARG}" == [0-9]* ]] || die "-X should be an integer\n"
      minstatus="${OPTARG}"
      ;;
    # Debug
    d)
      debug="${OPTARG}"
      ;;
    # Debug file
    D)
      debug_file="${OPTARG}"
      ;;
    # Help
    h)
      opthelp
      ;;
    *)
      opthelp
      ;;
  esac
done
shift $((OPTIND-1))

debug 3 "append [${append:-}]"
debug 3 "bell [${bell:-}]"
debug 3 "cmd [${cmd:-}]"
debug 3 "noclear [${noclear:-}]"
debug 3 "datefmt [${datefmt:-}]"
debug 3 "hosts [${hosts[@]:-}]"
debug 3 "down [${down:-}]"
debug 3 "up [${up:-}]"
debug 3 "pause [${pause:-}]"
debug 3 "retries [${retries:-}]"
debug 5 "usereturn [${usereturn:-}]"
debug 3 "config [${config:-}]"
debug 3 "changefile [${changefile:-}]"
debug 3 "maxwidth [${maxwidth:-}]"
debug 3 "minwidth [${minwidth:-}]"
debug 3 "minstatus [${minstatus:-}]"
debug 3 "debug [${debug:-}]"
debug 3 "debug_file [${debug_file:-}]"

date="date +${datefmt:-}"

# Make sure hosts were given
if [[ -z "${hosts[@]:-}" ]] ; then
  opthelp "No hosts defined\n"
fi

# Get config data
if ! source "$config" 2>/dev/null ; then
  die "Config file [$config] does not exist\n"
fi

# Check for the command
if ! type "${cmd%% *}" >/dev/null 2>&1 ; then
  die "[$cmd] does not exist\n"
fi

# Eval command line variable if it exists
if [[ -n "${exports:-}" ]] ; then
  eval "$exported" 2>/dev/null
fi

# Main loop
runcount=0
declare -a olddata=()
while true ; do
  debug 6 "Main loop"
  sdate=$($date)
  ((runcount++))
  declare host_print_ret=""

  # Try to define a line that spans the terminal
  # doing this here allows for terminal resize between runs
  if type tput >/dev/null 2>&1 ; then
    oldcols="${cols:-}"
    cols=$(tput cols)
    debug 6 "Terminal cols: old [${oldcols:-}] new [${cols:-}]"
    # Determine if we should print a new line - this shouldn't happen
    # often and these checks are faster than 'eval printf'
    if [[ -z "${oldcols:-}" || -z "${cols:-}" || -z "${line:-}" ||
      "${oldcols:-}" != "${cols:-}" ]] ; 
    then
      eval printf -v line '%.0s-' {1..$cols}
      debug 7 "Different terminal width - new horizontal rule"
    fi
  else
    if [[ -n "${line:-}" ]] ; then
      printf -v line '%.0s-' {1..80}
      debug 6 "Tput command not found - 80 col horizontal rule"
    fi
  fi

  declare -a data=()

  # Main loop to gather data
  for tmphost in "${hosts[@]}" ; do
    eval_host "${tmphost:-}"
  done

  # Output
  if [[ -z "${noclear:-}" ]] ; then
    clear
  fi
  echo -ne "Start time: $sdate\n"
  echo -ne "$host_print_ret"
  echo -n "End time: $($date) "

  # Write to a state change file
  if [[ -n "${changefile:-}" ]] ; then
    if [[ -n "${data[@]:-}" && -n "${olddata[@]:-}" && 
      "${data[@]:-}" != "${olddata[@]:-}" ]] ; then
      # Counter is odd numbered
      for (( count = 1 ; count < "${#data[@]}" ; count += 2 )) ; do
        if [[ "${data[$count]:-}" != "${olddata[$count]:-}" ]] ; then
          echo "[$($date)] [${data[$count -1]}] " \
            "${data[$count]:-}" >> "$changefile"
        fi
      done
    fi
    olddata="${data[@]:-}"
  fi

  # Limit runs
  if [[ -n "${runs:-}" ]] && (( runcount >= runs )) ; then
    break
  fi

  # Clean up
  unset host_print_ret
  sleep "$pause"
done

